Source/WebKit/WebProcess/GPU/webrtc/AudioMediaStreamTrackRendererInternalUnitManager.cpp

```diff
@@ -43,20 +43,30 @@
 #include <mach/mach_time.h>
 #include <wtf/Deque.h>
 #include <wtf/Identified.h>
+#include <wtf/RefCountedAndCanMakeWeakPtr.h>
 #include <wtf/TZoneMallocInlines.h>
 
 namespace WebKit {
 
-class AudioMediaStreamTrackRendererInternalUnitManagerProxy final : public WebCore::AudioMediaStreamTrackRendererInternalUnit, public CanMakeWeakPtr<AudioMediaStreamTrackRendererInternalUnitManagerProxy>, public Identified<AudioMediaStreamTrackRendererInternalUnitIdentifier> {
+class AudioMediaStreamTrackRendererInternalUnitManagerProxy final : public WebCore::AudioMediaStreamTrackRendererInternalUnit, public RefCountedAndCanMakeWeakPtr<AudioMediaStreamTrackRendererInternalUnitManagerProxy>, public Identified<AudioMediaStreamTrackRendererInternalUnitIdentifier> {
     WTF_MAKE_TZONE_ALLOCATED(AudioMediaStreamTrackRendererInternalUnitManagerProxy);
 public:
-    explicit AudioMediaStreamTrackRendererInternalUnitManagerProxy(Client&);
+    static Ref<AudioMediaStreamTrackRendererInternalUnitManagerProxy> create(Client& client)
+    {
+        return adoptRef(*new AudioMediaStreamTrackRendererInternalUnitManagerProxy(client));
+    }
+
     ~AudioMediaStreamTrackRendererInternalUnitManagerProxy();
 
+    void ref() const final { RefCounted::ref(); }
+    void deref() const final { RefCounted::deref(); }
+
     enum class IsClosed : bool { No, Yes };
     void reset(IsClosed);
 
 private:
+    explicit AudioMediaStreamTrackRendererInternalUnitManagerProxy(Client&);
+
```

Comment:
Same bug as a few days ago, `Client& m_client;` below is definitely not safe anymore now that you're extending lifetime.

Reply:
Ah you're totally right. In the newest revision I made `Client` inherit from `CanMakeWeakPtr` so that we can have a `WeakPtr` to it--but I'm not sure if I did it the right way. Of it's two derived classes, one of them is singleton, so I made it's ref/deref do nothing. But because it inherits from `CanMakeWeakPtr<itself>`, I had to use `USING_CAN_MAKE_WEAKPTR` and make `Client` be inherited publicly instead of privately. I'm not sure how this impacts the fact that it is also marked with `WeakPtrFactoryInitialization::Eager`.